File: data/src/main/java/org/cryptomator/data/cloud/crypto/Cryptors.java
package org.cryptomator.data.cloud.crypto;

import org.cryptomator.cryptolib.api.Cryptor;
import org.cryptomator.domain.Vault;
import org.cryptomator.domain.exception.MissingCryptorException;
import org.cryptomator.util.Optional;
import org.cryptomator.util.Supplier;

import java.util.Iterator;
import java.util.Map;
import java.util.concurrent.ConcurrentHashMap;
import java.util.concurrent.ConcurrentMap;

public abstract class Cryptors {

	Cryptors() {
	}

	public abstract boolean isEmpty();

	public abstract int size();

	public abstract Supplier<Cryptor> get(Vault vault);

	public abstract Optional<Cryptor> remove(Vault vault);

	public abstract boolean putIfAbsent(Vault vault, Cryptor cryptor);

	public static class Delegating extends Cryptors {

		private final Cryptors.Default fallback = new Cryptors.Default();

		private volatile Cryptors.Default delegate;

		public synchronized void setDelegate(Cryptors.Default delegate) {
			delegate.putAll(fallback.cryptors);
			this.delegate = delegate;
		}

		public synchronized void removeDelegate() {
			fallback.putAll(delegate.cryptors);
			this.delegate = null;
		}

		@Override
		public synchronized boolean isEmpty() {
			return delegate().isEmpty();
		}

		@Override
		public synchronized int size() {
			return delegate().size();
		}

		@Override
		public synchronized Supplier<Cryptor> get(Vault vault) {
			return delegate().get(vault);
		}

		@Override
		public synchronized Optional<Cryptor> remove(Vault vault) {
			return delegate().remove(vault);
		}

		@Override
		public synchronized boolean putIfAbsent(Vault vault, Cryptor cryptor) {
			return delegate().putIfAbsent(vault, cryptor);
		}

		private synchronized Cryptors delegate() {
			if (delegate == null) {
				return fallback;
			} else {
				return delegate;
			}
		}

	}

	public static class Default extends Cryptors {

		private final ConcurrentMap<Vault, Cryptor> cryptors = new ConcurrentHashMap<>();

		private Runnable onChangeListener = () -> {
		};

		public boolean isEmpty() {
			return cryptors.isEmpty();
		}

		public int size() {
			return cryptors.size();
		}

		public Supplier<Cryptor> get(final Vault vault) {
			return () -> {
				Cryptor cryptor = cryptors.get(vault);
				if (cryptor == null) {
					throw new MissingCryptorException();
				} else {
					return cryptor;
				}
			};
		}

		public Optional<Cryptor> remove(Vault vault) {
			Optional<Cryptor> result = Optional.ofNullable(cryptors.remove(vault));
			if (result.isPresent()) {
				onChangeListener.run();
			}
			return result;
		}

		public boolean putIfAbsent(Vault vault, Cryptor cryptor) {
			if (cryptors.putIfAbsent(vault, cryptor) == null) {
				onChangeListener.run();
				return true;
			} else {
				return false;
			}
		}

		public void setOnChangeListener(Runnable onChangeListener) {
			this.onChangeListener = onChangeListener;
		}

		public void putAll(Map<Vault, Cryptor> cryptors) {
			this.cryptors.putAll(cryptors);
			onChangeListener.run();
		}

		public void destroyAll() {
			while (!isEmpty()) {
				Iterator<Cryptor> cryptorIterator = cryptors.values().iterator();
				while (cryptorIterator.hasNext()) {
					cryptorIterator.next().destroy();
					cryptorIterator.remove();
				}
			}
			onChangeListener.run();
		}
	}

}

File: data/src/main/java/org/cryptomator/data/cloud/crypto/Cryptors.kt
package org.cryptomator.data.cloud.crypto

import com.google.common.base.Optional
import org.cryptomator.cryptolib.api.Cryptor
import org.cryptomator.domain.Vault
import org.cryptomator.domain.exception.MissingCryptorException
import java.util.concurrent.ConcurrentHashMap
import java.util.concurrent.ConcurrentMap
import java.util.function.Supplier

abstract class Cryptors internal constructor() {

	abstract fun isEmpty(): Boolean

	abstract fun size(): Int

	abstract operator fun get(vault: Vault): Supplier<Cryptor>

	abstract fun remove(vault: Vault): Optional<Cryptor>

	abstract fun putIfAbsent(vault: Vault, cryptor: Cryptor): Boolean

	class Delegating : Cryptors() {

		private val fallback = Default()

		@Volatile
		private var delegate: Default? = null

		@Synchronized
		fun setDelegate(delegate: Default) {
			delegate.putAll(fallback.cryptors)
			this.delegate = delegate
		}

		@Synchronized
		fun removeDelegate() {
			delegate?.let {
				fallback.putAll(it.cryptors)
			}.also { delegate = null }
		}

		@Synchronized
		override fun isEmpty(): Boolean {
			return delegate().isEmpty()
		}

		@Synchronized
		override fun size(): Int {
			return delegate().size()
		}

		@Synchronized
		override fun get(vault: Vault): Supplier<Cryptor> {
			return delegate()[vault]
		}

		@Synchronized
		override fun remove(vault: Vault): Optional<Cryptor> {
			return delegate().remove(vault)
		}

		@Synchronized
		override fun putIfAbsent(vault: Vault, cryptor: Cryptor): Boolean {
			return delegate().putIfAbsent(vault, cryptor)
		}

		@Synchronized
		private fun delegate(): Cryptors {
			return delegate ?: fallback
		}
	}

	class Default : Cryptors() {

		val cryptors: ConcurrentMap<Vault, Cryptor> = ConcurrentHashMap()

		private var onChangeListener = Runnable {}

		override fun isEmpty(): Boolean {
			return cryptors.isEmpty()
		}

		override fun size(): Int {
			return cryptors.size
		}

		override fun get(vault: Vault): Supplier<Cryptor> {
			return Supplier {
				cryptors[vault] ?: throw MissingCryptorException()
			}
		}

		override fun remove(vault: Vault): Optional<Cryptor> {
			val result = Optional.fromNullable(cryptors.remove(vault))
			if (result.isPresent) {
				onChangeListener.run()
			}
			return result
		}

		override fun putIfAbsent(vault: Vault, cryptor: Cryptor): Boolean {
			return if (cryptors.putIfAbsent(vault, cryptor) == null) {
				onChangeListener.run()
				true
			} else {
				false
			}
		}

		fun setOnChangeListener(onChangeListener: Runnable) {
			this.onChangeListener = onChangeListener
		}

		fun putAll(cryptors: Map<Vault, Cryptor>) {
			this.cryptors.putAll(cryptors)
			onChangeListener.run()
		}

		fun destroyAll() {
			while (!isEmpty()) {
				val cryptorIterator = cryptors.values.iterator()
				while (cryptorIterator.hasNext()) {
					cryptorIterator.next().destroy()
					cryptorIterator.remove()
				}
			}
			onChangeListener.run()
		}
	}
}
